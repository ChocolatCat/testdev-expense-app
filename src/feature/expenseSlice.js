import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import db, {app} from '../firebase/firebase';
import { ref, push, get, remove, update } from 'firebase/database';

//Thunk syntax for async functions
export const addExpense = createAsyncThunk('expenses/addExpense', 
async ({expense, uid}, thunkAPI) => {
    const {description = "No description", note = "No note", amount = 0, createdAt = 0} = expense;
    const expenseData = {description, note, amount, createdAt};
    const newExpense = push(ref(db, `users/${uid}/expenses/`), expenseData);
    return {id: newExpense.key, ...expenseData};
});

export const removeExpense = createAsyncThunk('expenses/removeExpense', 
async ({id, uid}, thunkAPI) => {
    await remove(ref(db, `users/${uid}/expenses/${id}`));
});

export const editExpense = createAsyncThunk('expenses/editExpense', 
async (expense, thunkAPI) => {
    await update(ref(db, `users/${expense.uid}/expenses/${expense.id}`), {
        ...expense.updates
    });
});

export const setExpenses = createAsyncThunk('expenses/setExpenses', 
async (uid, thunkAPI) => {
    const expenses = [];
    //we fetch values once. We use the await keyword to wait for the result
    //Really weird but doesn't work onValue
    await get(ref(db, `users/${uid}/expenses`)).then((snapshot) => {
        snapshot.forEach((childSnap) => {
            expenses.push({
                id:childSnap.key, 
                ...childSnap.val()
            });
        });
    });
    return expenses;
});

//Redux toolkit syntax
//Name defines the name in the store, initial state initializes the value and reducers are the actions that can manipulate the value
//Actions are defined by types, using the name as the first thing then a / to define the action. ie. counter/increment
//Reducers are pure functions - output is only determined by the input
//They never change state or action

export const expenseSlice = createSlice({
    name: 'expenses',
    initialState: [],
    reducers: {
    },
    extraReducers: (builder) => {
        builder.addCase(addExpense.fulfilled, (state, { payload }) => {
            state.push(payload);
        }),
        builder.addCase(addExpense.rejected, (state, { payload }) => {
            console.log('Tasukete!');
        }),
        builder.addCase(setExpenses.pending, (state, {payload}) => {
            console.log('Fetching Data...');
        }),
        builder.addCase(setExpenses.fulfilled, (state, {payload}) => {
            console.log('Done...');
            return payload;
        }),
        builder.addCase(setExpenses.rejected, (state, { payload }) => {
            console.log('Fetching Failed...');
            return [];
        }),
        builder.addCase(removeExpense.fulfilled, (state, { payload }) => {
            console.log("Removed");
        }),
        builder.addCase(removeExpense.rejected, (state, { payload }) => {
            console.log("Failed to Remove");
        }),
        builder.addCase(editExpense.fulfilled, (state, { payload }) => {
            console.log("Edited");
        }),
        builder.addCase(editExpense.rejected, (state, { payload }) => {
            console.log("Failed to Edit");
        })
    }
});

//export const { editExpense } = expenseSlice.actions;

export default expenseSlice.reducer;