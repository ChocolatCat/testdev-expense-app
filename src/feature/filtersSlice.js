import { createSlice } from '@reduxjs/toolkit'
import { startOfMonth, lastDayOfMonth } from 'date-fns';

//Redux toolkit syntax
//Name defines the name in the store, initial state initializes the value and reducers are the actions that can manipulate the value
//Actions are defined by types, using the name as the first thing then a / to define the action. ie. counter/increment
//Reducers are pure functions - output is only determined by the input
//They never change state or action

const filtersSlice = createSlice({
    name: 'filters',
    initialState: {
        text: '',
        sortBy: 'date',
        startDate: Date.parse(startOfMonth(Date.now())),
        endDate: Date.parse(lastDayOfMonth(Date.now()))
    },
    reducers: {
        setTextFilter: (state, action) => {
            state.text = action.payload ? action.payload : "";
        },
        sortByDate: (state) => {
            state.sortBy = 'date';
        },
        sortByAmount: (state) => {
            state.sortBy = 'amount';
        },
        setStartDate: (state, action) => {
            state.startDate = action.payload;
        },
        setEndDate: (state, action) => {
            state.endDate = action.payload;
        }
    }
});

export const {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} = filtersSlice.actions;

export default filtersSlice.reducer;