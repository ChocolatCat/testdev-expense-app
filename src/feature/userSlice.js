import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import {auth, googleAuthProvider} from '../firebase/firebase';
import { signInWithRedirect, signOut } from 'firebase/auth';

export const login = createAsyncThunk('user/login', 
async (arg, thunkAPI) => {
    signInWithRedirect(auth, googleAuthProvider).then((result) => {
        return true;
    });
});

export const logout = createAsyncThunk('user/logout', 
async (arg, thunkAPI) => {
    signOut(auth).then(() => {
        return false;
    });
});

const userSlice = createSlice({
    name: 'user',
    initialState: {
        isLogged: false,
        uid: undefined
    },
    reducers: {
        setLogStatus(state, {payload}) {
            state.isLogged = payload;
        },
        setUid(state, {payload}){
            state.uid = payload;
        }
    },
    extraReducers: (builder) => {
        builder.addCase(login.fulfilled, (state, {payload}) => {
            state.isLogged = payload;
        }),
        builder.addCase(login.rejected, (state, {payload}) => {
            state.isLogged = payload;
        });
    }
});

export const { setLogStatus, setUid } = userSlice.actions;

export default userSlice.reducer;