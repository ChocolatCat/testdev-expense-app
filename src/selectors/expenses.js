//import {createSelector} from '@reduxjs/toolkit';
import {isAfter, isBefore, isSameDay} from 'date-fns';

const selectExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = expense.createdAt;
        const startDateMatch = startDate ?  isAfter(createdAtMoment, startDate) || isSameDay(startDate, createdAtMoment) : true;
        const endDateMatch = endDate ? isBefore(createdAtMoment, endDate) || isSameDay(endDate, createdAtMoment) : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt > b.createdAt ? 1 : -1
        }
        if(sortBy === 'amount'){
            return a.amount > b.amount ? 1 : -1;
        }
    });  
};

const selectExpenseById = (expenses, id) => {
    return expenses.find( (expense) => {
        return expense.id === id;
    });
};

const selectExpensesTotal = (expenses) => {
    return expenses
        .map( (expense) => expense.amount )
        .reduce( (sum, value) => sum+value, 0);
}

export {selectExpenses, selectExpenseById, selectExpensesTotal};