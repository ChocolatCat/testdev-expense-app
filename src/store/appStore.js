import {configureStore} from '@reduxjs/toolkit';
import expenseReducer from '../feature/expenseSlice';
import filtersReducer from '../feature/filtersSlice';

export default () => {
    const store = configureStore({
        reducer:{
            expenses: expenseReducer,
            filters: filtersReducer
        }
    });
    return store;
}