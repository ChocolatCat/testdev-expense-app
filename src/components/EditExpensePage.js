import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {useParams, useNavigate} from 'react-router-dom';
import { selectExpenseById } from '../selectors/expenses';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../feature/expenseSlice'

//We use useParams to retrieve the parameters from the URL. We need a dynamic route using :param in the router.

const EditExpensePage = () => {
    const dispatch = useDispatch();
    const {id} = useParams();
    const navigate = useNavigate();
    const expense = useSelector(state => selectExpenseById(state.expenses, id));
    return (
        <div>
            <ExpenseForm 
                expense={expense}
                onSubmit={(expense) => {
                    dispatch(editExpense({
                        id: id,
                        updates: expense
                    }));
                    navigate('/');
                }}
            />
            <button onClick={()=>{
                dispatch(removeExpense(id));
                navigate('/');
            }}>Remove</button>
        </div>
    );
};

export default EditExpensePage;