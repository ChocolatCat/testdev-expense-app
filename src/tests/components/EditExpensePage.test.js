/** @jest-environment jsdom */

import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import {Provider} from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import EditExpensePage from '../../components/EditExpensePage';
import appStore from '../../store/appStore';
import {expenses} from '../fixtures/expenses';
import {addExpense} from '../../feature/expenseSlice';

//Mocks nanoid module because it's imported and used
jest.mock("nanoid", () => {
  return { nanoid: () => "1234" };
});

//mock navigate
const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
    useParams: () => ({
        id: '1234',
    })
}));

//We reset mocked navigate
beforeEach(() => {
  mockedUsedNavigate.mockReset();
});

jest.useFakeTimers().setSystemTime(new Date(2022, 7, 10, 8));


it('Should render without crashing', () => {
    //Must be in a provider to have a store. Must be in a router to use navigate
    const store = appStore();
    store.dispatch(addExpense(expenses[0]));
    const {asFragment} = render(
        <Provider store={store}>
            <BrowserRouter>
                <EditExpensePage />
            </BrowserRouter>
        </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});

it('Should submit to edit expense', () => {
    const store = appStore();
    store.dispatch(addExpense(expenses[0]));
    const {getByRole, getByPlaceholderText} = render(
        <Provider store={store}>
            <BrowserRouter>
                <EditExpensePage onSubmit={mockedUsedNavigate} />
            </BrowserRouter>
        </Provider>
    );
    fireEvent.change(getByPlaceholderText('Description'), {target: {value:"Renta"}});
    fireEvent.submit(getByRole('form'));
    expect(mockedUsedNavigate).toBeCalledTimes(1);
    expect(store.getState().expenses[0].description).toBe("Renta");
});

it('Should handle remove expense', () => {
    const store = appStore();
    store.dispatch(addExpense(expenses[0]));
    const {getByRole, getByText} = render(
        <Provider store={store}>
            <BrowserRouter>
                <EditExpensePage onSubmit={mockedUsedNavigate} />
            </BrowserRouter>
        </Provider>
    );
    fireEvent.click(getByText('Remove'));
    expect(mockedUsedNavigate).toBeCalledTimes(1);
    expect(store.getState().expenses).toEqual([]);
});