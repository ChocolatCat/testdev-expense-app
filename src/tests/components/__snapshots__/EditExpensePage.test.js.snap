// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render without crashing 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form
        name="expenseForm"
      >
        <input
          placeholder="Description"
          type="text"
          value="Gum"
        />
        <input
          placeholder="Amount"
          type="text"
          value="10"
        />
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              type="text"
              value="10/08/2022"
            />
          </div>
        </div>
        <textarea
          placeholder="Add a note to this expense! (optional)"
        />
        <button
          name="submit"
        >
          Add Expense
        </button>
      </form>
    </div>
    <button>
      Remove
    </button>
  </div>
</DocumentFragment>
`;
